File: backend/src/gamedeck/schemas/analytics.py
"""Analytics API response contracts."""

from datetime import datetime
from enum import StrEnum

from pydantic import BaseModel

from gamedeck.schemas.session import SessionResponse


class TimeBucket(StrEnum):
    DAY = "day"
    WEEK = "week"
    MONTH = "month"


class DistributionDimension(StrEnum):
    WEEKDAY = "weekday"
    HOUR = "hour"


class AnalyticsSummary(BaseModel):
    total_seconds: int
    session_count: int
    average_session_seconds: int
    longest_session_seconds: int


class GamePlaytime(BaseModel):
    game_id: int
    game_title: str
    total_seconds: int
    session_count: int


class SeriesPoint(BaseModel):
    bucket_start: datetime
    label: str
    total_seconds: int


class DistributionPoint(BaseModel):
    key: int
    label: str
    total_seconds: int


class PlaytimeResponse(BaseModel):
    from_at: datetime
    to_at: datetime
    time_zone: str
    bucket: TimeBucket
    summary: AnalyticsSummary
    series: list[SeriesPoint]
    games: list[GamePlaytime]


class DistributionResponse(BaseModel):
    from_at: datetime
    to_at: datetime
    time_zone: str
    dimension: DistributionDimension
    buckets: list[DistributionPoint]


class DashboardResponse(BaseModel):
    at: datetime
    time_zone: str
    today_seconds: int
    week_seconds: int
    month_seconds: int
    lifetime: AnalyticsSummary
    top_game: GamePlaytime | None
    current_sessions: list[SessionResponse]
    recent_sessions: list[SessionResponse]
    daily_series: list[SeriesPoint]


class GameAnalyticsResponse(BaseModel):
    game_id: int
    game_title: str
    from_at: datetime
    to_at: datetime
    time_zone: str
    summary: AnalyticsSummary
    daily_series: list[SeriesPoint]

File: frontend/src/types/analytics.ts
import type { GameSession } from '@/types/session'

export type TimeBucket = 'day' | 'week' | 'month'
export type DistributionDimension = 'weekday' | 'hour'

export interface AnalyticsSummary {
  total_seconds: number
  session_count: number
  average_session_seconds: number
  longest_session_seconds: number
}

export interface GamePlaytime {
  game_id: number
  game_title: string
  total_seconds: number
  session_count: number
}

export interface SeriesPoint {
  bucket_start: string
  label: string
  total_seconds: number
}

export interface DistributionPoint {
  key: number
  label: string
  total_seconds: number
}

export interface DashboardAnalytics {
  at: string
  time_zone: string
  today_seconds: number
  week_seconds: number
  month_seconds: number
  lifetime: AnalyticsSummary
  top_game: GamePlaytime | null
  current_sessions: GameSession[]
  recent_sessions: GameSession[]
  daily_series: SeriesPoint[]
}

export interface PlaytimeAnalytics {
  from_at: string
  to_at: string
  time_zone: string
  bucket: TimeBucket
  summary: AnalyticsSummary
  series: SeriesPoint[]
  games: GamePlaytime[]
}

export interface DistributionAnalytics {
  from_at: string
  to_at: string
  time_zone: string
  dimension: DistributionDimension
  buckets: DistributionPoint[]
}

export interface GameAnalytics {
  game_id: number
  game_title: string
  from_at: string
  to_at: string
  time_zone: string
  summary: AnalyticsSummary
  daily_series: SeriesPoint[]
}
